lib/sprintf.h

```diff
@@ -0,0 +1,101 @@
+/*
+ * SPDX-FileCopyrightText: 2023, Alejandro Colomar <alx@kernel.org>
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+
+#ifndef SHADOW_INCLUDE_LIB_SPRINTF_H_
+#define SHADOW_INCLUDE_LIB_SPRINTF_H_
+
+
+#include <config.h>
+
+#include <stdarg.h>
+#include <stddef.h>
+#include <stdio.h>
+
+#include "alloc.h"
+#include "defines.h"
+
+
+#if !defined(HAVE_ASPRINTF)
+format_attr(printf, 2, 3)
+inline int asprintf(char **restrict s, const char *restrict fmt, ...);
+format_attr(printf, 2, 0)
+inline int vasprintf(char **restrict s, const char *restrict fmt, va_list ap);
+#endif
+
+format_attr(printf, 2, 3)
+inline int xasprintf(char **restrict s, const char *restrict fmt, ...);
+format_attr(printf, 2, 0)
+inline int xvasprintf(char **restrict s, const char *restrict fmt, va_list ap);
+
+
+#if !defined(HAVE_ASPRINTF)
+inline int
+asprintf(char **restrict s, const char *restrict fmt, ...)
+{
+	int      len;
+	va_list  ap;
+
+	va_start(ap, fmt);
+	len = vasprintf(s, fmt, ap);
+	va_end(ap);
+
+	return len;
+}
+#endif
+
+
+#if !defined(HAVE_ASPRINTF)
+inline int
+vasprintf(char **restrict s, const char *restrict fmt, va_list ap)
```

Comment:
What systems does shadow support that lack vasprintf?

Reply:
I was wondering the same.. I had a quick look and saw musl and FreeBSD have it, then I thought at that point I should check with you to ask.

We could save a lot of dual implementions (which need testing too - ideally in CI, given it's a non-common codepath) if we're clearer about what we support 🤔

Reply:
Weee!

Reply:
> Probably none. Are we ready to claim support only for GNU / BSD /musl systems? vasprintf(3) is a GNU/BSD/musl extension.
> 
> Cc: @hallyn , @ikerexxe

No, I don't know of anyone using another libc, but I'd rather not lock them out if they exist.

Reply:
> We could add our own definition of vasprintf(3) if anyone needs it

That's fine.

Note that overall, given the risk for regressions,